recirq/qml_lfe/learn_dynamics_q.py

```diff
@@ -0,0 +1,238 @@
+import cirq_google
+import numpy as np
+from . import circuit_blocks
+
+from absl import app
+from absl import flags
+from absl import logging
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_integer("n", None, "System size (total qubits == 2 * n).")
+flags.DEFINE_integer("depth", None, "Circuit depth (block-wise).")
+flags.DEFINE_integer(
+    "n_data",
+    20,
+    "Number of circuits generated for each class (total circuits == 2 * n_data).",
+)
+
+flags.DEFINE_integer(
+    "batch_size",
+    5,
+    "Number of circuits to send over the wire per batch (value does not affect results).",
+)
+
+flags.DEFINE_integer(
+    "n_shots", 2000, "Number of measurements to draw from each individual circuit."
+)
+
+flags.DEFINE_string(
+    "save_dir", "./data", "Path to save experiment data (must already exist)."
+)
+
+flags.DEFINE_bool("use_engine", False, "Whether or not to use quantum engine.")
+
+
+def _flatten_circuit(circuit: cirq.Circuit) -> cirq.Circuit:
+    """Pack operations in circuit to the left as far as possible."""
+    return cirq.Circuit([op for mom in circuit for op in mom])
+
+
+def _build_circuit(
+    qubit_pairs: List[List[cirq.Qid]], use_tsym: bool, depth: int
+) -> cirq.Circuit:
+    """Generate twocopy 1D tsym or scrambling circuit with given depth on qubit_pairs."""
+    inter_gen = circuit_blocks.scrambling_block
+    if use_tsym:
+        # Use tsym gates.
+        inter_gen = circuit_blocks.tsym_block
+
+    # Random source for circuits.
+    random_source = np.random.uniform(0, 4, size=(depth * len(qubit_pairs), 2))
+
+    # Create initial bell pair circuit.
+    ret_circuit = cirq.Circuit(
+        [circuit_blocks.bell_pair_block(pair) for pair in qubit_pairs]
+    )
+
+    # Add tsym or scrambling block on System A qubits.
+    ret_circuit += circuit_blocks.block_1d_circuit(
+        [qbpair[0] for qbpair in qubit_pairs], depth, inter_gen, random_source
+    )
+
+    # Swap system A and system B qubits.
+    ret_circuit += [circuit_blocks.swap_block(pair) for pair in qubit_pairs]
+
+    # Again run tsym or scrambling block on newly swapped qubits.
+    ret_circuit += circuit_blocks.block_1d_circuit(
+        [qbpair[0] for qbpair in qubit_pairs], depth, inter_gen, random_source
+    )
+
+    # Un bell pair.
+    ret_circuit += [circuit_blocks.un_bell_pair_block(pair) for pair in qubit_pairs]
+
+    # Merge single qubit gates together and add measurements.
+    cirq.merge_single_qubit_gates_into_phxz(ret_circuit)
+    cirq.DropEmptyMoments().optimize_circuit(circuit=ret_circuit)
+    ret_circuit = _flatten_circuit(ret_circuit)
+
+    for i, qubit in enumerate([item for sublist in qubit_pairs for item in sublist]):
+        ret_circuit += cirq.measure(qubit, key="q{}".format(i))
+
+    cirq.SynchronizeTerminalMeasurements().optimize_circuit(circuit=ret_circuit)
+    logging.debug(
+        f"Generated a new circuit w/ tsym={use_tsym} and depth {len(ret_circuit)}"
+    )
+    return ret_circuit
+
+
+def _engine_sim_workaround(
```

Comment:
This is also a duplicated function.  Maybe it could be in a common file.

Reply:
Also split.